Mortar with materials

lindsayad opened the pull request:
@recuero I'm putting this up as a draft, mostly so you're aware that I'm working on it. All I can say about the current working state is that it compiles. I'm going to be adding moose tests in the coming day(s)

lindsayad committed:
Add MortarInterface

lindsayad committed:
Build map from lower-d to higher-d subdomains for mortar

lindsayad committed:
Allow material reinit and swap from all SubProblem derived classes

lindsayad committed:
Reinit materials for mortar

Leverages new abstract virtuals in SubProblem for reiniting
and swapping (via sentinels) materials.

Closes #14821

lindsayad committed:
Simple test of materials use in mortar

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/478724/) on e9a6254 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14872/site)

This comment will be updated on new commits.

lindsayad committed:
Set required mat props for mortar constraints

Do this by looping over MaterialInterface::getMatPropDepedencies

lindsayad committed:
clang-format v9 changes in FEProblemBase

moosebuild commented:
Job [Mac Test](https://civet.inl.gov/job/480104/) on 209fb2d : invalidated by @lindsayad

?

lindsayad committed:
Revert "Allow material reinit and swap from all SubProblem derived classes"

This reverts commit 95a5cb5ed5f740630eb75444fb13cde22afb3487.

lindsayad committed:
Use _fe_problem to reinit higher dimensional elements

I choose to do this for a couple of reasons:

1. If using _fe_problem then I don't have to change the API
of SwapBackSentinel.
2. This is much safer. This allows the MortarConstraints to use
material properties  coming from both displaced and undisplaced
materials. It's hard for me to imagine many scenarios where a user
would want to use an undisplaced material with a displaced constraint
or visa versa, but one example would be if using a
GenericConstantMaterial whose property  values are totally unrelated to
whether the mesh is displaced. It would be odd to require the user to
match the displaced setting of the constraints in this case. There is a
drawback here though, and that is speed. With this change we're going to
calculate variable values on both undisplaced and displaced meshes
regardless of whether they're needed. There's probably some checks we
can do in future commits that can limit computations done to
computations that  are needed. One idea I had was to do undisplaced and
displaced mortar constraints at the same time because then we could
potentially cut down on reinits by a factor of two...but this doesn't
really fit well into the current framework because we're looping over
mortar mesh elements (which will have fundamentally different
structures depending on whether the parent mesh is displaced or not)

recuero commented on framework/include/interfaces/MortarInterface.h:
```diff
@@ -0,0 +1,76 @@
+ *       the mortar interface's problem
+ */
+class MortarInterface
+{
```
`MortarInterface` makes the code cleaner 👍

recuero commented on test/tests/mortar/displaced-gap-conductance-2d-non-conforming/gap-conductance.i:
```diff
@@ -99,6 +99,16 @@
+  [constant]
+    type = GenericConstantMaterial
+    prop_names = 'gap_conductance'
+    prop_values = '.03'
```
Logic should work on non-constant Material Properties created internally (e.g. Mortar constraint initialization), right? Where are these projections happening in this PR?

lindsayad replied:
I'm not quite sure what you're asking but i evaluate slave and master face material properties [here](https://github.com/idaholab/moose/pull/14872/files#diff-1526b1ec72736a09e386eb5a57012066R162) and [here](https://github.com/idaholab/moose/pull/14872/files#diff-1526b1ec72736a09e386eb5a57012066R191) respectively.

recuero reviewed (commented):
@lindsayad Thanks for creating this capability. Being able to use `MaterialProperty` within `MortarConstraint` will allow us to model thermal problems with existing material models.

lindsayad commented on framework/include/constraints/MortarData.h:
```diff
@@ -112,4 +121,8 @@ class MortarData
+
+  /// Map from lower dimensional subdomain ids to corresponding higher simensional subdomain ids
+  /// (e.g. the ids of the interior parents)
+  std::unordered_map<SubdomainID, std::set<SubdomainID>> _lower_d_sub_to_higher_d_subs;
```
I dunno. Maybe in theory. I'll remove this commit because I didn't end up using that code

lindsayad commented on framework/include/interfaces/MortarInterface.h:
```diff
@@ -0,0 +1,76 @@
+/**
+ * An interface for accessing mortar mesh data
+ *
+ * \note moi is shorthand for mortar interface, so \p _moi_problem indicates
```
This is a pattern inherited from other interfaces like `MaterialPropertyInterface`. The prefix is so that there is no name collision with however the problem is stored by the `MooseObject`, I believe

lindsayad commented on framework/include/loops/ComputeMortarFunctor.h:
```diff
@@ -34,7 +35,9 @@ class ComputeMortarFunctor
       const AutomaticMortarGeneration & amg,
-      SubProblem & subproblem);
+      SubProblem & subproblem,
+      FEProblemBase & _fe_problem,
```
Not necessarily. You do not know in the general sense whether the mortar constraints are acting on the undisplaced or displaced problem

fdkong commented on framework/include/constraints/MortarData.h:
```diff
@@ -12,14 +12,17 @@
 class SubProblem;
 
-class MortarData
+class MortarData : public libMesh::ParallelObject
```
Just curious why this does not derive from MooseObject?

lindsayad replied:
I just didn't see the need yet to go through and create `InputParameters` and all that jazz. If this starts taking more construction arguments, then I can see changing its inheritance to `MooseObject`

fdkong commented on framework/include/interfaces/MortarInterface.h:
```diff
@@ -0,0 +1,76 @@
+  /**
+   * Returns the master lower dimensional subdomain id
+   */
+  SubdomainID masterSubdomain() const { return _master_subdomain_id; }
```
Need slaveSubdomain as well?

lindsayad replied:
The only reason this API exists is for subdomain coverage checks in the `ConstraintWarehouse`. I'm able to get the slave subdomain through the associated lagrange multiplier variable, but because it doesn't live on the master subdomain I had to add this API to `MortarData` to get the master

fdkong commented:
I am fine with this PR. @lindsayad
